src/smoke.hh

```diff
@@ -1,16 +1,39 @@
 #include <nanogui/nanogui.h>
 #include <list>
 #include "emittor.hh"
+#include "navier_stoke_solver.hh"
 
 class Smoke
 {
 
 public:
     Smoke(){};
+    Smoke(int width, int height, int grid_width, int grid_height)
+    {
+        this->width = width;
+        this->height = height;
+        this->grid_width = grid_width;
+        this->grid_height = grid_height;
+        this->nsp = NavierStokeSolver();
+    };
+
     ~Smoke(){};
 
+    NavierStokeSolver nsp;
     std::list<Particle> particles;
 
-    void generateParticles(const Emittor& emittor, int num_particles);
+    int width;
+    int height;
+    int depth;
+    int grid_width;
+    int grid_height;
+    int grid_depth;
+
+    void generateParticles(Emittor emittor, int num_particles);
     void update(double delta_t);
+    void build_spatial_map();
```

Comment:
use `const Emitter& emitter`
Try pass by reference when ever possible for classes